File: packages/mobile/android/app/src/main/java/id/homebase/lib/core/file/DriveFileUploadProvider.java
package id.homebase.lib.core.file;

import static id.homebase.lib.core.crypto.CryptoUtil.cbcEncrypt;
import static id.homebase.lib.core.crypto.CryptoUtil.encryptKeyHeader;
import static id.homebase.lib.core.crypto.CryptoUtil.encryptMetaData;
import static id.homebase.lib.core.crypto.CryptoUtil.stringToByteArray;
import static id.homebase.lib.core.file.types.KeyHeaderGenerator.generateKeyHeader;
import static id.homebase.lib.core.file.types.KeyHeaderGenerator.getRandom16ByteArray;

import android.util.Log;

import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.PropertyAccessor;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;

import org.json.JSONObject;

import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

import id.homebase.lib.core.DotYouClient;
import id.homebase.lib.core.HttpClientOptions;
import id.homebase.lib.core.crypto.CryptoUtil;
import id.homebase.lib.core.file.types.BadRequestUploadResult;
import id.homebase.lib.core.file.types.EncryptedKeyHeader;
import id.homebase.lib.core.file.types.KeyHeader;
import id.homebase.lib.core.file.types.StreamRequestBody;
import id.homebase.lib.core.file.types.SuccessfullUploadResult;
import id.homebase.lib.core.file.types.UploadFileMetadata;
import id.homebase.lib.core.file.types.UploadInstructionSet;
import id.homebase.lib.core.file.types.UploadManifest;
import id.homebase.lib.core.file.types.UploadPayloadDescriptor;
import id.homebase.lib.core.file.types.UploadResult;
import id.homebase.lib.core.file.types.UploadThumbnailDescriptor;
import id.homebase.lib.core.file.types.payloadorthumbnailbase.PayloadBase;
import id.homebase.lib.core.file.types.payloadorthumbnailbase.PayloadOrThumbnailBase;
import id.homebase.lib.core.file.types.payloadorthumbnailbase.PayloadOrThumbnailFile;
import id.homebase.lib.core.file.types.payloadorthumbnailbase.PayloadOrThumbnailStream;
import id.homebase.lib.core.file.types.payloadorthumbnailbase.ThumbnailBase;
import kotlin.NotImplementedError;
import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;
import okhttp3.ResponseBody;

public class DriveFileUploadProvider {
    public static UploadResult uploadFile(
            DotYouClient dotYouClient,
            UploadInstructionSet instructions,
            UploadFileMetadata<String> metadata,
            List<PayloadBase> payloads,
            List<ThumbnailBase> thumbnails,
            boolean encrypt
    ) throws Exception {
        // Debug information
        if (isDebug()) {
            Log.v(null, "request: " + dotYouClient.getEndpoint() + "/drive/files/upload" +
                    " instructions: " + instructions +
                    " metadata: " + metadata +
                    " payloads: " + payloads +
                    " thumbnails: " + thumbnails);
        }

        // Force isEncrypted on the metadata to match the encrypt flag
        metadata.setIsEncrypted(encrypt);

        KeyHeader keyHeader = encrypt ? generateKeyHeader() : null;

        UploadManifest manifest = buildManifest(payloads, thumbnails, encrypt);
        instructions.setManifest(manifest);
        instructions.setTransferIv(instructions.getTransferIv() != null ?
                instructions.getTransferIv() : getRandom16ByteArray());

        // Build package
        byte[] encryptedDescriptor = buildDescriptor(
                dotYouClient, keyHeader, instructions, metadata
        );

        // Upload
        MultipartBody data = buildFormData(
                instructions, encryptedDescriptor, payloads, thumbnails, keyHeader, manifest
        );

        return pureUpload(dotYouClient, data);
    }

    public static boolean isDebug() {
        // Return your debug flag
        return true;
    }

    private static UploadManifest buildManifest(
            List<PayloadBase> payloads,
            List<ThumbnailBase> thumbnails,
            boolean generateIv
    ) {
        List<UploadPayloadDescriptor> payloadDescriptors = payloads.stream()
                .map(payload -> {
                    List<UploadThumbnailDescriptor> relatedThumbnails = thumbnails.stream()
                            .filter(thumb -> thumb.getKey().equals(payload.getKey()))
                            .map(thumb -> new UploadThumbnailDescriptor(
                                    thumb.getKey() + thumb.getPixelWidth(),
                                    thumb.getPixelHeight(),
                                    thumb.getPixelWidth()))
                            .collect(Collectors.toList());

                    return new UploadPayloadDescriptor(
                            payload.getKey(),
                            payload.getDescriptorContent(),
                            relatedThumbnails,
                            payload.getPreviewThumbnail(),
                            generateIv ? getRandom16ByteArray() : null
                    );
                })
                .collect(Collectors.toList());

        return new UploadManifest(payloadDescriptors);
    }

    private static byte[] buildDescriptor(
            DotYouClient dotYouClient,
            KeyHeader keyHeader,
            UploadInstructionSet instructions,
            UploadFileMetadata<String> metadata
    ) throws Exception {

        if (instructions.getTransferIv() == null) {
            throw new IllegalArgumentException("Transfer IV is required");
        }

        return encryptWithSharedSecret(dotYouClient,
                new DescriptorData(encryptKeyHeader(dotYouClient, keyHeader, instructions.getTransferIv()), encryptMetaData(metadata, keyHeader)),
                instructions.getTransferIv());
    }

    private static byte[] encryptWithSharedSecret(
            DotYouClient dotYouClient,
            DescriptorData data,
            byte[] iv) throws Exception {

        byte[] ss = dotYouClient.getSharedSecret();
        if (ss == null) {
            throw new IllegalStateException("Attempting to encrypt but missing the shared secret");
        }

        byte[] content = data.toJsonString().getBytes();

        return cbcEncrypt(content, iv, ss);
    }

    private static MultipartBody buildFormData(
            UploadInstructionSet instructions,
            byte[] encryptedDescriptor,
            List<PayloadBase> payloads,
            List<ThumbnailBase> thumbnails,
            KeyHeader keyHeader,
            UploadManifest manifest
    )
            throws Exception {
        MultipartBody.Builder builder = new MultipartBody.Builder()
                .setType(MultipartBody.FORM);

        builder.addFormDataPart("instructions", null, RequestBody.create(stringToByteArray(instructions.toJsonString()), MediaType.parse("application/octet-stream")));

        if (encryptedDescriptor != null) {
            builder.addFormDataPart("metaData", null, RequestBody.create(encryptedDescriptor, MediaType.parse("application/octet-stream")));
        }

        if (payloads != null) {
            for (PayloadBase payload : payloads) {
                RequestBody payloadBody;

                if (keyHeader == null) {
                    payloadBody = getFileOrStreamRequestBody(payload);
                } else {
                    java.io.File payloadFile = payload instanceof PayloadOrThumbnailFile ? ((PayloadOrThumbnailFile) payload).getPayload() : null;
                    InputStream payloadStream = payload instanceof PayloadOrThumbnailStream ? ((PayloadOrThumbnailStream) payload).getInputStream() : null;

                    ByteArrayOutputStream encryptedPayload = payloadFile != null ? CryptoUtil.encryptWithKeyheader(payloadFile, getUpdatedKeyHeader(keyHeader, manifest, payload.getKey())) : CryptoUtil.encryptWithKeyheader(payloadStream, getUpdatedKeyHeader(keyHeader, manifest, payload.getKey()));
                    payloadBody = new StreamRequestBody(encryptedPayload, MediaType.parse(payload.getContentType()));
                }

                builder.addFormDataPart("payload", payload.getKey(), payloadBody);
            }
        }

        if (thumbnails != null) {
            for (ThumbnailBase thumb : thumbnails) {
                RequestBody payloadBody;

                if (keyHeader == null) {
                    payloadBody = getFileOrStreamRequestBody(thumb);
                } else {
                    java.io.File thumbFile = thumb instanceof PayloadOrThumbnailFile ? ((PayloadOrThumbnailFile) thumb).getPayload() : null;
                    InputStream thumbStream = thumb instanceof PayloadOrThumbnailStream ? ((PayloadOrThumbnailStream) thumb).getInputStream() : null;

                    ByteArrayOutputStream encryptedPayload = thumbFile != null ? CryptoUtil.encryptWithKeyheader(thumbFile, getUpdatedKeyHeader(keyHeader, manifest, thumb.getKey())) : CryptoUtil.encryptWithKeyheader(thumbStream, getUpdatedKeyHeader(keyHeader, manifest, thumb.getKey()));
                    payloadBody = new StreamRequestBody(encryptedPayload, MediaType.parse(thumb.getContentType()));
                }

                builder.addFormDataPart("thumbnail", thumb.getKey() + thumb.getPixelWidth(), payloadBody);
            }
        }

        return builder.build();

    }

    private static RequestBody getFileOrStreamRequestBody(PayloadOrThumbnailBase payloadOrThumbnailBase) {
        if (payloadOrThumbnailBase instanceof PayloadOrThumbnailStream) {
            return new StreamRequestBody(((PayloadOrThumbnailStream) payloadOrThumbnailBase).getOutputStream(), MediaType.parse(payloadOrThumbnailBase.getContentType()));
        } else if (payloadOrThumbnailBase instanceof PayloadOrThumbnailFile) {
            return RequestBody.create(((PayloadOrThumbnailFile) payloadOrThumbnailBase).getPayload(), MediaType.parse(payloadOrThumbnailBase.getContentType()));
        }
        return null;
    }

    private static KeyHeader getUpdatedKeyHeader(KeyHeader keyHeader, UploadManifest manifest, String payloadKey) {
        if (manifest != null && manifest.payloadDescriptors() != null) {
            for (UploadPayloadDescriptor descriptor : manifest.payloadDescriptors()) {
                if (descriptor.payloadKey().equals(payloadKey)) {
                    if (descriptor.iv() != null) {
                        return new KeyHeader(descriptor.iv(), keyHeader.aesKey());
                    }
                }
            }
        }
        return keyHeader;
    }

    private static UploadResult pureUpload(
            DotYouClient dotYouClient,
            MultipartBody data
    ) {
        Request request = new Request.Builder()
                .url(dotYouClient.getEndpoint() + "/drive/files/upload")
                .post(data)
                .build();

        try (
                Response response = dotYouClient.createHttpClient(new HttpClientOptions(true)).newCall(request).execute();
                ResponseBody body = response.body()) {

            if (body != null) {
                String jsonData = body.string();
                JSONObject uploadResultObject = new JSONObject(jsonData);

                if (response.isSuccessful()) {
                    return new SuccessfullUploadResult(uploadResultObject);
                } else if (response.code() == 400) {
                    return new BadRequestUploadResult(uploadResultObject);
                }

                // TODO
                Log.e(null, "Error: " + jsonData);
            }

            // Do something if the body is `null`

        } catch (Exception e) {
            // TODO
            Log.e(null, "Error: " + e.getMessage() + Arrays.toString(e.getStackTrace()));
        }

        throw new NotImplementedError();
    }

    private static class DescriptorData {
        private final EncryptedKeyHeader encryptedKeyHeader;
        private final UploadFileMetadata<String> fileMetadata;

        public DescriptorData(EncryptedKeyHeader encryptedKeyHeader, UploadFileMetadata<String> fileMetadata) {
            this.encryptedKeyHeader = encryptedKeyHeader;
            this.fileMetadata = fileMetadata;
        }

        public String toJsonString() throws JsonProcessingException {
            ObjectMapper objectMapper = new ObjectMapper();
            objectMapper.setVisibility(PropertyAccessor.FIELD, JsonAutoDetect.Visibility.ANY);

            return objectMapper.writeValueAsString(this);
        }
    }
}

File: packages/mobile/ios/core/DriveFileUploadProvider.swift
//
//  DriveFileUploadProvider.swift
//  HomebasePhotos
//
//  Created by Stef Coenen on 19/06/2024.
//

import Foundation
import CryptoKit
import Alamofire

extension DriveFileUploadProvider {
  static func uploadFile(dotYouClient: DotYouClient, instructions: UploadInstructionSet, metadata: UploadFileMetadata<String>, payloads: [PayloadBase], thumbnails: [ThumbnailBase], encrypt: Bool, completion: @escaping UploadCompletionHandler) throws -> Void  {
    print("[SyncWorker] Upload");

    // Implement the uploadFile logic
    var mutableMetadata = metadata
    mutableMetadata.setIsEncrypted(encrypt)

    let keyHeader = encrypt ? generateKeyHeader() : nil
    let manifest = buildManifest(payloads: payloads, thumbnails: thumbnails, generateIv: encrypt)

    let updatedInstructions = UploadInstructionSet(storageOptions: instructions.storageOptions, transitOptions: instructions.transitOptions, transferIv: instructions.transferIv ?? getRandom16ByteArray(), manifest: manifest)

    let encryptedDescriptor = try buildDescriptor(dotYouClient: dotYouClient, keyHeader: keyHeader, instructions: updatedInstructions, metadata: mutableMetadata)
    let formData = try buildFormData(instructions: updatedInstructions, encryptedDescriptor: encryptedDescriptor, payloads: payloads, thumbnails: thumbnails, keyHeader: keyHeader, manifest: manifest)

    return try pureUpload(dotYouClient: dotYouClient, data: formData, completion: completion)
  }

  static func buildManifest(payloads: [PayloadBase], thumbnails: [ThumbnailBase], generateIv: Bool) -> UploadManifest {
    let payloadDescriptors: [UploadPayloadDescriptor] = payloads.map { payload in
      let relatedThumbnails: [UploadThumbnailDescriptor] = thumbnails.filter { thumb in
        thumb.key == payload.key
      }.map { thumb in
        UploadThumbnailDescriptor(
          thumbnailKey: thumb.key + String(thumb.pixelWidth),
          pixelHeight: thumb.pixelHeight,
          pixelWidth: thumb.pixelWidth
        )
      }

      return UploadPayloadDescriptor(
        payloadKey: payload.key,
        descriptorContent: payload.descriptorContent ?? nil,
        thumbnails: relatedThumbnails,
        previewThumbnail: payload.previewThumbnail,
        iv: generateIv ? getRandom16ByteArray() : nil
      )
    }

    return UploadManifest(payloadDescriptors: payloadDescriptors)
  }

  static func buildDescriptor(dotYouClient: DotYouClient, keyHeader: KeyHeader?, instructions: UploadInstructionSet, metadata: UploadFileMetadata<String>) throws -> Data {
    guard let transferIv = instructions.transferIv else {
      throw NSError(domain: "DriveFileUploadProvider", code: 1, userInfo: [NSLocalizedDescriptionKey: "Transfer IV is required"])
    }

    let descriptorData = DescriptorData(encryptedKeyHeader: try CryptoUtil.encryptKeyHeader( dotYouClient: dotYouClient, keyHeader: keyHeader != nil ? keyHeader! : getEmptyKeyHeader(), transferIv: instructions.transferIv!), fileMetadata: metadata)
    let jsonString = descriptorData.toJsonString()
    let content = jsonString.data(using: .utf8)

    return try CryptoUtil.cbcEncrypt(data: content!, iv: transferIv, key: dotYouClient.sharedSecret!)
  }

  static func buildFormData(instructions: UploadInstructionSet, encryptedDescriptor: Data, payloads: [PayloadBase], thumbnails: [ThumbnailBase], keyHeader: KeyHeader?, manifest: UploadManifest) throws -> MultipartFormData {
    let formData = MultipartFormData()

    // Add instructions
    if let instructionsData = instructions.toJsonString().data(using: .utf8) {
      formData.append(instructionsData, withName: "instructions", fileName: "instructions.json", mimeType: "application/octet-stream")
    }

    // Add encrypted descriptor if available
    formData.append(encryptedDescriptor, withName: "metaData", fileName: "metadata.bin", mimeType: "application/octet-stream")

    // Add payloads and thumbnails as streams
    for payload in payloads {
      if let keyHeader = keyHeader {
        if let payloadFile = payload as? PayloadFile {
          let encryptedPayload = try CryptoUtil.encryptWithKeyHeader(file: payloadFile.filePath, keyHeader: CryptoUtil.getUpdatedKeyHeader(keyHeader: keyHeader, manifest: manifest, payloadKey: payload.key))
          formData.append(encryptedPayload.stream, withLength: encryptedPayload.count, name: "payload", fileName: payload.key, mimeType: payload.contentType)

        } else if let payloadStream = payload as? PayloadStream {
          let encryptedPayload = try CryptoUtil.encryptWithKeyHeader(stream: payloadStream.inputStream.stream, keyHeader: CryptoUtil.getUpdatedKeyHeader(keyHeader: keyHeader, manifest: manifest, payloadKey: payload.key))
          formData.append(encryptedPayload.stream, withLength: encryptedPayload.count, name: "payload", fileName: payload.key, mimeType: payload.contentType)
        } else {
          continue
        }
      } else {
        if let payloadFile = payload as? PayloadFile {
          if #available(iOS 16.0, *) {
            formData.append(InputStream(fileAtPath: payloadFile.filePath)!, withLength: try FileManager.default.attributesOfItem(atPath: payloadFile.filePath)[.size] as! UInt64, name: "payload", fileName: payload.key, mimeType: payload.contentType)
          } else {
            // Fallback on earlier versions
          }
        } else if let payloadStream = payload as? PayloadStream {
          formData.append(payloadStream.inputStream.stream, withLength: payloadStream.inputStream.count, name: "payload", fileName: payload.key, mimeType: payload.contentType)
        } else {
          continue
        }
      }
    }

    for thumb in thumbnails {
      if let keyHeader = keyHeader {
        if let thumbStream = thumb as? ThumbnailStream {
          let encryptedThumb = try CryptoUtil.encryptWithKeyHeader(stream: thumbStream.inputStream.stream, keyHeader: CryptoUtil.getUpdatedKeyHeader(keyHeader: keyHeader, manifest: manifest, payloadKey: thumb.key))
          formData.append(encryptedThumb.stream, withLength: encryptedThumb.count, name: "thumbnail", fileName: thumb.key + String(thumb.pixelWidth), mimeType: thumb.contentType);
        } else {
          continue
        }
      } else {
        if let thumbStream = thumb as? ThumbnailStream {
          formData.append(thumbStream.inputStream.stream, withLength: thumbStream.inputStream.count, name: "thumbnail", fileName: thumb.key + String(thumb.pixelWidth), mimeType: thumb.contentType);
        } else {
          continue
        }
      }
    }

    return formData
  }

  typealias UploadCompletionHandler = (Result<String?, Error>) -> Void
  static func pureUpload(dotYouClient: DotYouClient, data: MultipartFormData, completion: @escaping UploadCompletionHandler
) throws -> Void  {
    AF.upload(
      multipartFormData: data,
      to:dotYouClient.endpoint + "/drive/files/upload",
      usingThreshold: UInt64.init(),
      method: .post,
      headers: HTTPHeaders(dotYouClient.headers)
    )
    .validate(statusCode: 200..<401)
    .responseString { response in
      // Handle the response
      switch response.result {
      case .success(let data):
        let result = JsonParser.parseJson(jsonString: data);
        
        if(response.response?.statusCode == 200 ){
          completion(.success(data))
        } else if(result?["errorCode"] as! String == "existingFileWithUniqueId"){
          completion(.success(data))
        } else {
          completion(.failure(NSError(domain: "DriveFileUploadProvider", code: 1, userInfo: [NSLocalizedDescriptionKey: "Upload failed with a bad request"])))
        }
      case .failure(let error):
          completion(.failure(error))
      }
    }
  }

  static func generateKeyHeader() -> KeyHeader {
    // Generate key header
    return KeyHeader(iv: getRandom16ByteArray(), aesKey: getRandom16ByteArray())
  }

  static func getRandom16ByteArray() -> Data {
    // Generate random 16-byte array
    return SymmetricKey(size: .bits128).withUnsafeBytes { Data($0) }
  }
  
  static func getEmptyKeyHeader() -> KeyHeader {
    return KeyHeader(iv: Data(repeating: 0, count: 16), aesKey: Data(repeating: 0, count: 16) )
  }
}

extension Data {
  mutating func append(_ string: String) {
    if let data = string.data(using: .utf8) {
      append(data)
    }
  }

  init(reading input: InputStream) throws {
    self.init()
    input.open()
    defer { input.close() }

    let bufferSize = 1024
    var buffer = [UInt8](repeating: 0, count: bufferSize)

    while input.hasBytesAvailable {
      let read = input.read(&buffer, maxLength: bufferSize)
      if read > 0 {
        append(buffer, count: read)
      } else if read < 0 {
        throw input.streamError ?? NSError(domain: NSPOSIXErrorDomain, code: Int(errno), userInfo: nil)
      } else {
        break
      }
    }
  }
}


// URLSession synchronous data task extension
extension URLSession {
  func synchronousDataTask(with request: URLRequest) -> (Data?, URLResponse?, Error?) {
    var data: Data?
    var response: URLResponse?
    var error: Error?

    let semaphore = DispatchSemaphore(value: 0)

    let dataTask = self.dataTask(with: request) {
      data = $0
      response = $1
      error = $2
      semaphore.signal()
    }

    dataTask.resume()
    semaphore.wait()

    return (data, response, error)
  }
}

struct TargetDrive  :Codable {
  let alias: String
  let type: String
}

struct AccessControlList : Codable {
  let type: SecurityGroupType
}

enum SecurityGroupType :Codable {
  case owner
}

struct DriveFileUploadProvider {
}


struct StorageOptions  :Codable {
  let drive: TargetDrive
}

struct UploadManifest  :Codable {
  let payloadDescriptors: [UploadPayloadDescriptor]
}

struct UploadPayloadDescriptor  :Codable {
  let payloadKey: String
  let descriptorContent: String?
  let thumbnails: [UploadThumbnailDescriptor]
  let previewThumbnail: EmbeddedThumb?
  let iv: Data?
}

struct UploadThumbnailDescriptor  :Codable {
  let thumbnailKey: String
  let pixelHeight: Int
  let pixelWidth: Int
}


struct UploadInstructionSet :Codable {
  let storageOptions: StorageOptions
  var transitOptions: TransitOptions?
  var transferIv: Data?
  var manifest: UploadManifest?

  func toJsonString() -> String {
    let encoder = JSONEncoder()
    encoder.outputFormatting = .prettyPrinted

    // Custom data encoding strategy for `Data`
    encoder.dataEncodingStrategy = .base64

    do {
      let jsonData = try encoder.encode(self)
      if let jsonString = String(data: jsonData, encoding: .utf8) {
        return jsonString
      }
    } catch {
      print("Failed to convert to JSON string: \(error)")
    }
    return "{}"
  }
}

struct TransitOptions  :Codable {
  // Not used ATM
}

struct UploadAppFileMetaData<T: Codable> :Codable {
  let uniqueId: String
  let tags: [String]
  let fileType: Int
  let dataType: Int
  let userDate: Int64
  let groupId: String?
  let archivalStatus: Int
  var content:  String
  var previewThumbnail: EmbeddedThumb?
}

struct UploadFileMetadata<T: Codable>: Codable {
  let allowDistribution: Bool
  var isEncrypted: Bool
  let acl: AccessControlList
  var appData: UploadAppFileMetaData<T>
  let referencedFile: String?

  mutating func setIsEncrypted(_ encrypt: Bool) {
    self.isEncrypted = encrypt
  }
}


class UploadResult {}
class SuccessfulUploadResult: UploadResult {}
class BadRequestUploadResult: UploadResult {
  var errorCode: String? = nil
}


struct KeyHeader {
  var iv: Data
  var aesKey: Data
}

struct EncryptedKeyHeader : Codable {
  var iv: Data
  var encryptedAesKey: Data
  var encryptionVersion: Int
  var type: Int
}

struct DescriptorData : Codable {
  var encryptedKeyHeader: EncryptedKeyHeader
  var fileMetadata: UploadFileMetadata<String>

  func toJsonString() -> String {
    let encoder = JSONEncoder()
    encoder.outputFormatting = .prettyPrinted

    // Custom data encoding strategy for `Data`
    encoder.dataEncodingStrategy = .base64

    do {
      let jsonData = try encoder.encode(self)
      if let jsonString = String(data: jsonData, encoding: .utf8) {
        return jsonString
      }
    } catch {
      print("Failed to convert to JSON string: \(error)")
    }
    return "{}"
  }
}

class JsonParser {
  static func parseJson(jsonString: String) -> [String: Any]? {
    if let jsonData = jsonString.data(using: .utf8) {
      do {
        let jsonObject = try JSONSerialization.jsonObject(with: jsonData, options: [])
        if let dictionary = jsonObject as? [String: Any] {
            return dictionary
        }
      } catch {
        print("Failed to parse JSON: \(error.localizedDescription)")
      }
    }
    
    return nil
  }
  
 static func traverseJSON(_ json: Any, parentKey: String = "") {
      if let dictionary = json as? [String: Any] {
          for (key, value) in dictionary {
              let fullKey = parentKey.isEmpty ? key : "\(parentKey).\(key)"
              traverseJSON(value, parentKey: fullKey)
          }
      } else if let array = json as? [Any] {
          for (index, value) in array.enumerated() {
              let fullKey = "\(parentKey)[\(index)]"
              traverseJSON(value, parentKey: fullKey)
          }
      } else {
          print("\(parentKey): \(json)")
      }
  }

}
